BUG: handle nan values in DataFrame.update when overwrite=False (#15593)

pcluo opened the pull request:
BUG: handle nan values in DataFrame.update when overwrite=False (#15593)

add nan test for DataFrame.update

update whatsnew v0.20.2

 - [x] closes #15593 
closes #15617
 - [x] tests added / passed
 - [x] passes ``git diff upstream/master --name-only -- '*.py' | flake8 --diff``
 - [x] whatsnew entry

pcluo committed:
BUG: handle nan values in DataFrame.update when overwrite=False (#15593)

add nan test for DataFrame.update

update whatsnew v0.20.2

TomAugspurger commented on doc/source/whatsnew/v0.20.2.txt:
```diff
@@ -37,6 +37,7 @@ Bug Fixes
 ~~~~~~~~~
 
 - Bug in using ``pathlib.Path`` or ``py.path.local`` objects with io functions (:issue:`16291`)
+- Bug in using ``DataFrame.update`` when overwrite=False and the second value is nan
```
Could you add a ref to the issue ath the end here

    (:issue:`15593`)

TomAugspurger commented on pandas/tests/frame/test_combine_concat.py:
```diff
@@ -763,3 +763,13 @@ def test_concat_datetime_datetime64_frame(self):
+        df1 = DataFrame({'A': [1, None, 3], 'B': date_range('2000', periods=3)})
+        df2 = DataFrame({'A': [None, 2, 3]})
+        df1.update(df2, overwrite=False)
+        # it works!
```
could you make an exected dataframe and compare?

```python
expected = df1 = DataFrame({'A': [1.0, 2, 3], 'B': date_range('2000', periods=3)})
tm.assert_frame_equal(df1, expected)
```

pcluo commented:
yes, i just made the changes.

On Mon, May 22, 2017 at 3:58 PM, Tom Augspurger <notifications@github.com>
wrote:

> *@TomAugspurger* commented on this pull request.
> ------------------------------
>
> In doc/source/whatsnew/v0.20.2.txt
> <https://github.com/pandas-dev/pandas/pull/16430#discussion_r117831062>:
>
> > @@ -37,6 +37,7 @@ Bug Fixes
>  ~~~~~~~~~
>
>  - Bug in using ``pathlib.Path`` or ``py.path.local`` objects with io functions (:issue:`16291`)
> +- Bug in using ``DataFrame.update`` when overwrite=False and the second value is nan
>
> Could you add a ref to the issue ath the end here
>
> (:issue:`15593`)
>
> ------------------------------
>
> In pandas/tests/frame/test_combine_concat.py
> <https://github.com/pandas-dev/pandas/pull/16430#discussion_r117831818>:
>
> > @@ -763,3 +763,13 @@ def test_concat_datetime_datetime64_frame(self):
>
>          # it works!
>          pd.concat([df1, df2_obj])
> +
> +
> +class TestDataFrameUpdate(TestData):
> +
> +    def test_update_nan(self):
> +        # #15593 #15617
> +        df1 = DataFrame({'A': [1, None, 3], 'B': date_range('2000', periods=3)})
> +        df2 = DataFrame({'A': [None, 2, 3]})
> +        df1.update(df2, overwrite=False)
> +        # it works!
>
> could you make an exected dataframe and compare?
>
> expected = df1 = DataFrame({'A': [1.0, 2, 3], 'B': date_range('2000', periods=3)})
> tm.assert_frame_equal(df1, expected)
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/pandas-dev/pandas/pull/16430#pullrequestreview-39570384>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABSVsjx9erVVSzQBeyQlFJlbjjRXyqd3ks5r8ejbgaJpZM4Niywe>
> .
>

TomAugspurger commented on pandas/tests/frame/test_combine_concat.py:
```diff
@@ -763,3 +763,14 @@ def test_concat_datetime_datetime64_frame(self):
+        df2 = DataFrame({'A': [None, 2, 3]})
+        df1.update(df2, overwrite=False)
+        # it works!
+        tm.assert_frame_equal(df1, expected)
```
Can you add a second test with 

```python
df1 = DataFrame({'A': [1,None,3], 'B': date_range('2000', periods=3)})
df1.update(df2, overwrite=False)
```

just to make sure we cover the original issue? You can do it in this same test, and use the same `expected`

jreback commented on pandas/tests/frame/test_combine_concat.py:
```diff
@@ -763,3 +763,14 @@ def test_concat_datetime_datetime64_frame(self):
+
+    def test_update_nan(self):
+        # #15593 #15617
+        expected = df1 = DataFrame({'A': [1.0, 2, 3], 'B': date_range('2000', periods=3)})
```
do like this

```
df1 = DataFrame(....)
df2 = DataFrame(....)

expected = df1
result = df1.update(....)

tm.assert_frame_equal(result, expected)
```

TomAugspurger replied:
I think update is inplace? So expected = df1.copy()

jreback replied:
no it returns a new object (but has inplace arg)

jreback replied:
sorry, you are right @TomAugspurger . Hmm that is pretty bogus it should have a kwarg for this.

jreback replied:
https://github.com/pandas-dev/pandas/issues/10730

jreback commented on doc/source/whatsnew/v0.20.2.txt:
```diff
@@ -37,6 +37,7 @@ Bug Fixes
 ~~~~~~~~~
 
 - Bug in using ``pathlib.Path`` or ``py.path.local`` objects with io functions (:issue:`16291`)
+- Bug in using ``DataFrame.update`` when overwrite=False and the second value is nan (:issue:`15593`)
```
add the 2nd issue as well. 

Bug in ``DataFrame.update()`` with ``overwrite=False`` and ``NaN`` values

TomAugspurger commented:
@pcluo can you update with the requested changes? Thanks!

pcluo commented:
@TomAugspurger yes, plz check.

TomAugspurger commented on pandas/tests/frame/test_combine_concat.py:
```diff
@@ -763,3 +763,23 @@ def test_concat_datetime_datetime64_frame(self):
+        tm.assert_frame_equal(df1, expected)
+
+        # test 2
+        expected = df1 = DataFrame({'A': [1, None, 3], 'B': date_range('2000', periods=3)})
```
Sorry, I think this got lost in the conversation, but can you write this as:

```python
df1 = DataFrame({'A': [1, None, 3], 'B': date_range('2000', periods=3)})
df2 = DataFrame({'A': [None, 2, 3]})
expected = DataFrame({'A': [1, 2, 3], 'B': date_range('2000', periods=3)})
```

The one on line 773 can be a bit shorter, just 

```python
df1 = DataFrame({'A': [1.0, 2, 3], 'B': date_range('2000', periods=3)})
expected = df1.copy()
```

basically, since `.update` works inplace, the test will always be true no matter what happens. `expected` and `df1` are reference to the same object. It's like

```python
In [2]: a = b = {'x': 1}

In [3]: a['y'] = 2

In [4]: a == b
Out[4]: True
```

Which "shouldn't" be true (but is, since `a` and `b` are both references to the same object. Let me know if that makes sense.

pcluo commented:
@t

pcluo commented:
@TomAugspurger just made another attempt at this. Sorry the last conversation got lost when i squashed commits. I changed [1, 2, 3] in test 2 to [1.0, 2, 3] otherwise the dtypes are not equal, float64 vs int64.

jreback commented:
can you rebase on master

TomAugspurger commented:
>  I changed [1, 2, 3] in test 2 to [1.0, 2, 3] otherwise the dtypes are not equal, float64 vs int64.

Yeah, that sounds right.

If you rebase and force push, all the tests should pass.

```
git fetch upstream
git rebase upstream/master
git push -f origin master
```

FYI, typically you want to checkout a new branch for features / bug fixes. http://pandas.pydata.org/pandas-docs/stable/contributing.html#creating-a-branch Not a huge deal though.

pcluo commented:
it seems all tests passed now. thx @TomAugspurger  and will checkout a new branch next time.

TomAugspurger committed:
Restart TravisCI

TomAugspurger commented:
Looks like Travis isn't picking up your commits for some reason. I'll ask their support.

jreback commented:
thanks!

TomAugspurger commented:
FYI @pcluo TravisCI support thought that if you sign into travis with your github account then things should be good.
